File: TESTING/EIG/dcklse.f
      SUBROUTINE DCKLSE( NN, MVAL, PVAL, NVAL, NMATS, ISEED, THRESH, NMAX, A, AF, B, BF, X, WORK, RWORK, NIN, NOUT, INFO )

*  -- LAPACK test routine --
*  -- LAPACK is a software package provided by Univ. of Tennessee,    --
*  -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--

      // .. Scalar Arguments ..
      int                INFO, NIN, NMATS, NMAX, NN, NOUT;
      double             THRESH;
      // ..
      // .. Array Arguments ..
      int                ISEED( 4 ), MVAL( * ), NVAL( * ), PVAL( * );
      double             A( * ), AF( * ), B( * ), BF( * ), RWORK( * ), WORK( * ), X( * );
      // ..

*  =====================================================================

      // .. Parameters ..
      int                NTESTS;
      const              NTESTS = 7 ;
      int                NTYPES;
      const              NTYPES = 8 ;
      // ..
      // .. Local Scalars ..
      bool               FIRSTT;
      String             DISTA, DISTB, TYPE;
      String             PATH;
      int                I, IINFO, IK, IMAT, KLA, KLB, KUA, KUB, LDA, LDB, LWORK, M, MODEA, MODEB, N, NFAIL, NRUN, NT, P;
      double             ANORM, BNORM, CNDNMA, CNDNMB;
      // ..
      // .. Local Arrays ..
      bool               DOTYPE( NTYPES );
      double             RESULT( NTESTS );
      // ..
      // .. External Subroutines ..
      // EXTERNAL ALAHDG, ALAREQ, ALASUM, DLARHS, DLATB9, DLATMS, DLSETS
      // ..
      // .. Intrinsic Functions ..
      // INTRINSIC ABS, MAX
      // ..
      // .. Executable Statements ..

      // Initialize constants and the random number seed.

      PATH( 1: 3 ) = 'LSE'
      INFO = 0
      NRUN = 0
      NFAIL = 0
      FIRSTT = .TRUE.
      alareq(PATH, NMATS, DOTYPE, NTYPES, NIN, NOUT );
      LDA = NMAX
      LDB = NMAX
      LWORK = NMAX*NMAX

      // Check for valid input values.

      for (IK = 1; IK <= NN; IK++) { // 10
         M = MVAL( IK )
         P = PVAL( IK )
         N = NVAL( IK )
         if ( P.GT.N .OR. N.GT.M+P ) {
            if ( FIRSTT ) {
               WRITE( NOUT, FMT = * )
               FIRSTT = .FALSE.
            }
            WRITE( NOUT, FMT = 9997 )M, P, N
         }
   10 CONTINUE
      FIRSTT = .TRUE.

      // Do for each value of M in MVAL.

      for (IK = 1; IK <= NN; IK++) { // 40
         M = MVAL( IK )
         P = PVAL( IK )
         N = NVAL( IK )
         IF( P.GT.N .OR. N.GT.M+P ) GO TO 40

         for (IMAT = 1; IMAT <= NTYPES; IMAT++) { // 30

            // Do the tests only if DOTYPE( IMAT ) is true.

            IF( .NOT.DOTYPE( IMAT ) ) GO TO 30

            // Set up parameters with DLATB9 and generate test
            // matrices A and B with DLATMS.

            dlatb9(PATH, IMAT, M, P, N, TYPE, KLA, KUA, KLB, KUB, ANORM, BNORM, MODEA, MODEB, CNDNMA, CNDNMB, DISTA, DISTB );

            dlatms(M, N, DISTA, ISEED, TYPE, RWORK, MODEA, CNDNMA, ANORM, KLA, KUA, 'No packing', A, LDA, WORK, IINFO );
            if ( IINFO.NE.0 ) {
               WRITE( NOUT, FMT = 9999 )IINFO
               INFO = ABS( IINFO )
               GO TO 30
            }

            dlatms(P, N, DISTB, ISEED, TYPE, RWORK, MODEB, CNDNMB, BNORM, KLB, KUB, 'No packing', B, LDB, WORK, IINFO );
            if ( IINFO.NE.0 ) {
               WRITE( NOUT, FMT = 9999 )IINFO
               INFO = ABS( IINFO )
               GO TO 30
            }

            // Generate the right-hand sides C and D for the LSE.

            dlarhs('DGE', 'New solution', 'Upper', 'N', M, N, MAX( M-1, 0 ), MAX( N-1, 0 ), 1, A, LDA, X( 4*NMAX+1 ), MAX( N, 1 ), X, MAX( M, 1 ), ISEED, IINFO );

            dlarhs('DGE', 'Computed', 'Upper', 'N', P, N, MAX( P-1, 0 ), MAX( N-1, 0 ), 1, B, LDB, X( 4*NMAX+1 ), MAX( N, 1 ), X( 2*NMAX+1 ), MAX( P, 1 ), ISEED, IINFO );

            NT = 2

            dlsets(M, P, N, A, AF, LDA, B, BF, LDB, X, X( NMAX+1 ), X( 2*NMAX+1 ), X( 3*NMAX+1 ), X( 4*NMAX+1 ), WORK, LWORK, RWORK, RESULT( 1 ) );

            // Print information about the tests that did not
            // pass the threshold.

            for (I = 1; I <= NT; I++) { // 20
               if ( RESULT( I ).GE.THRESH ) {
                  if ( NFAIL.EQ.0 .AND. FIRSTT ) {
                     FIRSTT = .FALSE.
                     alahdg(NOUT, PATH );
                  }
                  WRITE( NOUT, FMT = 9998 )M, P, N, IMAT, I, RESULT( I )
                  NFAIL = NFAIL + 1
               }
   20       CONTINUE
            NRUN = NRUN + NT

   30    CONTINUE
   40 CONTINUE

      // Print a summary of the results.

      alasum(PATH, NOUT, NFAIL, NRUN, 0 );

 9999 FORMAT( ' DLATMS in DCKLSE   INFO = ', I5 )
 9998 FORMAT( ' M=', I4, ' P=', I4, ', N=', I4, ', type ', I2, ', test ', I2, ', ratio=', G13.6 )
 9997 FORMAT( ' *** Invalid input  for LSE:  M = ', I6, ', P = ', I6, ', N = ', I6, ';', / '     must satisfy P <= N <= P+M  ', '(this set of values will be skipped)' )
      RETURN

      // End of DCKLSE

      }

File: test/eig/dcklse.dart
// Copyright (c) 2024 Guilherme Lepsch. All rights reserved. Use of this
// source code is governed by a BSD-style license that can be found in the
// [LICENSE file](https://github.com/lepsch/dart_lapack/blob/main/LICENSE).

import 'dart:math';

import 'package:dart_lapack/src/box.dart';
import 'package:dart_lapack/src/format_specifiers_extensions.dart';
import 'package:dart_lapack/src/matrix.dart';
import 'package:dart_lapack/src/nio.dart';
import 'package:dart_lapack/src/range.dart';
import 'package:test/test.dart';

import '../lin/alasum.dart';
import '../matgen/dlatms.dart';
import '../test_driver.dart';
import 'alahdg.dart';
import 'alareq.dart';
import 'dlarhs.dart';
import 'dlatb9.dart';
import 'dlsets.dart';

Future<void> dcklse(
  final int NN,
  final Array<int> MVAL_,
  final Array<int> PVAL_,
  final Array<int> NVAL_,
  final int NMATS,
  final Array<int> ISEED_,
  final double THRESH,
  final int NMAX,
  final Array<double> A_,
  final Array<double> AF_,
  final Array<double> B_,
  final Array<double> BF_,
  final Array<double> X_,
  final Array<double> WORK_,
  final Array<double> RWORK_,
  final Nin NIN,
  final Nout NOUT,
  final Box<int> INFO,
  final TestDriver test,
  final String group,
) async {
// -- LAPACK test routine --
// -- LAPACK is a software package provided by Univ. of Tennessee,    --
// -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--
  final MVAL = MVAL_.having();
  final PVAL = PVAL_.having();
  final NVAL = NVAL_.having();
  final ISEED = ISEED_.having();
  final A = A_.having();
  final AF = AF_.having();
  final B = B_.having();
  final BF = BF_.having();
  final X = X_.having();
  final WORK = WORK_.having();
  final RWORK = RWORK_.having();
  const NTESTS = 7;
  const NTYPES = 8;
  final DOTYPE = Array<bool>(NTYPES);
  final RESULT = Array<double>(NTESTS);
  const PATH = 'LSE';

  // Initialize constants and the random number seed.

  INFO.value = 0;
  var NRUN = 0;
  var NFAIL = 0;
  await alareq(PATH, NMATS, DOTYPE, NTYPES, NIN, NOUT);
  final LDA = NMAX;
  final LDB = NMAX;
  final LWORK = NMAX * NMAX;

  test.group(group, () {
    var FIRSTT = true;

    // Check for valid input values.

    for (var IK = 1; IK <= NN; IK++) {
      final M = MVAL[IK];
      final P = PVAL[IK];
      final N = NVAL[IK];
      if (P > N || N > M + P) {
        if (FIRSTT) {
          NOUT.println();
          FIRSTT = false;
        }
        NOUT.println(
            ' *** Invalid input  for LSE:  M = ${M.i6}, P = ${P.i6}, N = ${N.i6};\n     must satisfy P <= N <= P+M  (this set of values will be skipped)');
      }
    }
    FIRSTT = true;

    // Do for each value of M in MVAL.

    for (final IK in 1.through(NN)) {
      final M = MVAL[IK];
      final P = PVAL[IK];
      final N = NVAL[IK];
      if (P > N || N > M + P) continue;

      for (final IMAT in 1.through(NTYPES)) {
        // Do the tests only if DOTYPE[ IMAT ] is true.
        final skip = !DOTYPE[IMAT];
        test('DCKLSE (M = $M, N = $N, P = $P TYPE = $IMAT)', () {
          final IINFO = Box(0);

          // Set up parameters with DLATB9 and generate test
          // matrices A and B with DLATMS.

          final (
            :TYPE,
            :KLA,
            :KUA,
            :KLB,
            :KUB,
            :ANORM,
            :BNORM,
            :MODEA,
            :MODEB,
            :CNDNMA,
            :CNDNMB,
            :DISTA,
            :DISTB
          ) = dlatb9(PATH, IMAT, M, P, N);

          dlatms(M, N, DISTA, ISEED, TYPE, RWORK, MODEA, CNDNMA, ANORM, KLA,
              KUA, 'No packing', A.asMatrix(LDA), LDA, WORK, IINFO);
          test.expect(IINFO.value, 0);
          if (IINFO.value != 0) {
            print9999(NOUT, IINFO.value);
            INFO.value = IINFO.value.abs();
            return;
          }

          dlatms(P, N, DISTB, ISEED, TYPE, RWORK, MODEB, CNDNMB, BNORM, KLB,
              KUB, 'No packing', B.asMatrix(LDB), LDB, WORK, IINFO);
          test.expect(IINFO.value, 0);
          if (IINFO.value != 0) {
            print9999(NOUT, IINFO.value);
            INFO.value = IINFO.value.abs();
            return;
          }

          // Generate the right-hand sides C and D for the LSE.
          var LD1 = max(N, 1), LD2 = max(M, 1);

          dlarhs(
              'DGE',
              'New solution',
              'Upper',
              'N',
              M,
              N,
              max(M - 1, 0),
              max(N - 1, 0),
              1,
              A.asMatrix(LDA),
              LDA,
              X(4 * NMAX + 1).asMatrix(LD1),
              LD1,
              X.asMatrix(LD2),
              LD2,
              ISEED,
              IINFO);

          LD1 = max(N, 1);
          LD2 = max(P, 1);
          dlarhs(
              'DGE',
              'Computed',
              'Upper',
              'N',
              P,
              N,
              max(P - 1, 0),
              max(N - 1, 0),
              1,
              B.asMatrix(LDB),
              LDB,
              X(4 * NMAX + 1).asMatrix(LD1),
              LD1,
              X(2 * NMAX + 1).asMatrix(LD2),
              LD2,
              ISEED,
              IINFO);

          const NT = 2;

          dlsets(
              M,
              P,
              N,
              A.asMatrix(LDA),
              AF.asMatrix(LDA),
              LDA,
              B.asMatrix(LDB),
              BF.asMatrix(LDB),
              LDB,
              X,
              X(NMAX + 1),
              X(2 * NMAX + 1),
              X(3 * NMAX + 1),
              X(4 * NMAX + 1),
              WORK,
              LWORK,
              RWORK,
              RESULT(1));

          // Print information about the tests that did not
          // pass the threshold.

          for (var I = 1; I <= NT; I++) {
            final reason =
                ' M=${M.i4} P=${P.i4}, N=${N.i4}, type ${IMAT.i2}, test ${I.i2}, ratio=${RESULT[I].g13_6}';
            test.expect(RESULT[I], lessThan(THRESH), reason: reason);
            if (RESULT[I] >= THRESH) {
              if (NFAIL == 0 && FIRSTT) {
                FIRSTT = false;
                alahdg(NOUT, PATH);
              }
              NOUT.println(reason);
              NFAIL++;
            }
          }
          NRUN += NT;
        }, skip: skip);
      }
    }
  });

  // Print a summary of the results.
  alasum(PATH, NOUT, NFAIL, NRUN, 0);
}

void print9999(final Nout nout, final int info) {
  nout.println(' DLATMS in DCKLSE   INFO = ${info.i5}');
}
